// @sanity/desk-tool/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _DeskTool = _interopRequireDefault(require("./DeskTool"));

var _viewColumnIcon = _interopRequireDefault(require("part:@sanity/base/view-column-icon"));

var _uuid = _interopRequireDefault(require("@sanity/uuid"));

var _router = require("part:@sanity/base/router");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = {
  router: (0, _router.route)('/:selectedType', [(0, _router.route)('/:action', [(0, _router.route)('/:selectedDocumentId')])]),

  canHandleIntent(intentName, params) {
    return intentName === 'edit' && params.id || intentName === 'create' && params.type;
  },

  getIntentState(intentName, params) {
    return {
      selectedType: params.type || '*',
      action: 'edit',
      selectedDocumentId: params.id || (0, _uuid.default)()
    };
  },

  title: 'Desk',
  name: 'desk',
  icon: _viewColumnIcon.default,
  component: _DeskTool.default
};
exports.default = _default;
