// @sanity/components/lib/panes/SplitPaneWrapper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class SplitPaneWrapper extends _react.default.Component {
  render() {
    return this.props.children || _react.default.createElement("div", null);
  }

}

exports.default = SplitPaneWrapper;

_defineProperty(SplitPaneWrapper, "propTypes", {
  minWidth: _propTypes.default.number,
  maxWidth: _propTypes.default.number,
  defaultWidth: _propTypes.default.number,
  children: _propTypes.default.node
});

_defineProperty(SplitPaneWrapper, "defaultProps", {
  minWidth: 0,
  maxWidth: 0,
  defaultWidth: 0
});
