// @sanity/desk-tool/lib/pane/DocumentsPaneMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _default = _interopRequireDefault(require("part:@sanity/components/menus/default"));

var _barsIcon = _interopRequireDefault(require("part:@sanity/base/bars-icon"));

var _thListIcon = _interopRequireDefault(require("part:@sanity/base/th-list-icon"));

var _thLargeIcon = _interopRequireDefault(require("part:@sanity/base/th-large-icon"));

var _plusCircleIcon = _interopRequireDefault(require("part:@sanity/base/plus-circle-icon"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const TEST_CARDS_AND_THUMBNAILS = false;
const LIST_VIEW_ITEMS = [{
  title: 'List',
  icon: _barsIcon.default,
  action: 'setListLayout',
  key: 'default',
  divider: true
}, {
  title: 'Details',
  icon: _thListIcon.default,
  action: 'setListLayout',
  key: 'detail'
}, // Disabled for now as it needs more work (virtual-list support, etc.)
TEST_CARDS_AND_THUMBNAILS && {
  title: 'Thumbnails (__DEV__ only)',
  icon: _thLargeIcon.default,
  action: 'setListLayout',
  key: 'media'
}, // Disabled for now as it needs more work (virtual-list support, etc.)
TEST_CARDS_AND_THUMBNAILS && {
  title: 'Cards (__DEV__ only)',
  icon: _thLargeIcon.default,
  action: 'setListLayout',
  key: 'card'
}, {
  title: 'Create new…',
  icon: _plusCircleIcon.default,
  action: 'createNew',
  divider: true
}].filter(Boolean);

const NULL_COMPONENT = () => null;

class DocumentsPaneMenu extends _react.default.PureComponent {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "handleMenuAction", item => {
      if (item.action === 'setListLayout') {
        this.props.onSetListLayout(item);
      }

      if (item.action === 'setOrdering') {
        this.props.onSetOrdering(item.ordering);
      }

      if (item.action === 'createNew') {
        this.props.onGoToCreateNew();
      }

      this.props.onMenuClose();
    });
  }

  render() {
    const orderingOptions = this.props.orderingOptions;
    const orderingItems = orderingOptions.map(orderingOption => ({
      title: orderingOption.title,
      icon: orderingOption.icon || NULL_COMPONENT,
      ordering: orderingOption,
      action: 'setOrdering',
      key: orderingOption.name
    })).concat(LIST_VIEW_ITEMS);
    return _react.default.createElement(_default.default, _extends({
      onAction: this.handleMenuAction,
      items: orderingItems,
      origin: "top-right"
    }, this.props));
  }

}

exports.default = DocumentsPaneMenu;

_defineProperty(DocumentsPaneMenu, "propTypes", {
  onSetListLayout: _propTypes.default.func,
  onSetOrdering: _propTypes.default.func,
  onGoToCreateNew: _propTypes.default.func,
  onMenuClose: _propTypes.default.func,
  orderingOptions: _propTypes.default.array
});
