// @sanity/desk-tool/lib/pane/TypePaneItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _TypePaneItem = _interopRequireDefault(require("./styles/TypePaneItem.css"));

var _router = require("part:@sanity/base/router");

var _reactInk = _interopRequireDefault(require("react-ink"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class TypePaneItem extends _react.default.Component {
  render() {
    const _this$props = this.props,
          selected = _this$props.selected,
          type = _this$props.type;
    return _react.default.createElement("div", {
      className: selected ? _TypePaneItem.default.selected : _TypePaneItem.default.item,
      key: document._id
    }, _react.default.createElement(_router.StateLink, {
      state: {
        selectedType: type.name
      },
      className: _TypePaneItem.default.link
    }, type.title, _react.default.createElement(_reactInk.default, {
      duration: 1000,
      opacity: 0.1,
      radius: 200
    })));
  }

}

exports.default = TypePaneItem;

_defineProperty(TypePaneItem, "propTypes", {
  type: _propTypes.default.shape({
    name: _propTypes.default.string,
    title: _propTypes.default.string,
    icon: _propTypes.default.element
  }),
  selected: _propTypes.default.bool
});
